### DISCUSSION TITLE
How to query nested collection properties?

### DISCUSSION THREAD

**Discussion by xenoterracide** [mikro-orm/mikro-orm#3902] [Q&A] [answered]
This is weird?

```ts
export class StorefrontProductVariationValue
  implements Tradable, StorefrontProductVariation
{

  @ValidateNested()
  @Property({ type: 'array' })
  readonly components: StorefrontComponent[] = [];
}

export type StorefrontComponent = DeepReadonly<StorefrontComponentValue>; // DeepReadonly is from ts-essentials

export class StorefrontComponentValue implements StorefrontComponent {
  @Property({ type: 'uuid' })
  readonly componentId: ComponentId;

  @Property({ type: 'boolean' })
  available = true;
```

but my query doesn't expose `componentId or available`

```ts
  constructor(
    @InjectRepository(StorefrontProductAggregate)
    private readonly dao: EntityRepository<StorefrontProductAggregate>,
  ) {}

  async findForAvailabilityChange(
    storefrontId: StorefrontId,
    componentId: ComponentId,
  ): Promise<readonly StorefrontProductAggregate[]> {
    return this.dao.find({
      storefrontId,
      variations: {
        components: {
          componentId, // compile error here, list doesn't show any props
        },
      },
    });
  }
 ```
 
singular and collection properties on variations show.

how can I query this? note: this is using mongodb.

**Comment by hayksaryan** [comment, marked as answer]
Take a look at this [section of the official documentation](https://mikro-orm.io/docs/entity-manager#searching-by-referenced-entity-fields).

You need to specify an additional populate option.
```
return this.dao.find({
      storefrontId,
      variations: {
        components: {
          componentId, // compile error here, list doesn't show any props
        },
      },
    }, { populate: ['variations.components']});

```

**Reply by B4nan** [reply]
This is not about population, he defines a scalar array property and expects it to somehow map things to entity instances - if it's supposed to be mapping to entities (or any classes in general), it needs to be a relation property (collection instance here) or an embedded array property. Since he mentioned he uses mongo, I'd say an embedded array would work the best, as due to the lack of join support in mongo, you can't do much when querying the relations in there.

I am not sure if the issue in OP is purely TS problem, maybe it is? Because he mentiones he is getting some values, if it would be filtering issue, there would be no return value at all.

If it is just TS problem, I'd suggest trying it without the `DeepReadonly` thing, and more importantly, sharing the full error message. 

But it feels more like the OP is trying to use some random classes (not discovered by the ORM) - decorating them with entity/embeddable decorators (e.g. `@Property`) won't have any effect unless the class is a discovered entity/embeddable. You won't get any class mapping on JSON properties without the use of embeddables (or custom mapped types, that's always an alternative).

**Reply by xenoterracide** [reply]
I think I had problems with all of the above... I also think part of it was I didn't realize I was getting multiple collections which I didn't want. But I also had that compile error... That's all been resolved now.
